File: controllers/segment_controller.py
from flask import request, jsonify
from models.segment import Segment
from models.project import Project
from config.database import get_db

def get_segments():
    """Obtener todos los segmentos"""
    try:
        print('📹 Obteniendo todos los segmentos')
        
        # Obtener base de datos
        db = get_db()
        
        # Obtener todos los segmentos
        segments = Segment.find_all(db)
        
        print(f'✅ Segmentos encontrados: {len(segments)}')
        
        # Convertir a formato de respuesta
        segments_data = [segment.to_response_dict() for segment in segments]
        
        response = {
            'success': True,
            'message': 'Segmentos obtenidos exitosamente',
            'data': {
                'segments': segments_data,
                'count': len(segments_data)
            }
        }
        
        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)
        
    except Exception as error:
        print('💥 Error al obtener segmentos:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al obtener segmentos'
        }), 500

def get_segment(segment_id):
    """Obtener un segmento por ID"""
    try:
        print(f'📹 Obteniendo segmento con ID: {segment_id}')
        
        # Obtener base de datos
        db = get_db()
        
        # Buscar segmento
        segment = Segment.find_by_id(db, segment_id)
        
        if not segment:
            print(f'❌ Segmento no encontrado: {segment_id}')
            return jsonify({
                'success': False,
                'message': 'Segmento no encontrado'
            }), 404
        
        print(f'✅ Segmento encontrado: {segment_id}')
        
        response = {
            'success': True,
            'message': 'Segmento obtenido exitosamente',
            'data': {
                'segment': segment.to_response_dict()
            }
        }
        
        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)
        
    except Exception as error:
        print('💥 Error al obtener segmento:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al obtener segmento'
        }), 500

def get_segments_by_project(project_id):
    """Obtener segmentos por proyecto"""
    try:
        print(f'📹 Obteniendo segmentos del proyecto: {project_id}')
        print(f'📋 Tipo de project_id: {type(project_id)}')
        
        # Validar project_id
        if not project_id:
            print('❌ project_id es None o vacío')
            return jsonify({
                'success': False,
                'message': 'ID de proyecto requerido'
            }), 400
        
        # Obtener base de datos
        db = get_db()
        print(f'🗄️ Base de datos conectada: {db.name}')
        
        # Verificar que el proyecto existe
        print(f'🔍 Verificando existencia del proyecto: {project_id}')
        project = Project.find_by_id(db, project_id)
        if not project:
            print(f'❌ Proyecto no encontrado: {project_id}')
            return jsonify({
                'success': False,
                'message': 'Proyecto no encontrado'
            }), 404
        
        print(f'✅ Proyecto encontrado: {project_id}')
        
        # Obtener segmentos del proyecto
        print(f'🔍 Buscando segmentos para proyecto: {project_id}')
        segments = Segment.find_by_project(db, project_id)
        
        print(f'✅ Segmentos encontrados: {len(segments)}')
        
        # Convertir a formato de respuesta
        print('🔄 Convirtiendo segmentos a formato de respuesta...')
        segments_data = []
        for i, segment in enumerate(segments):
            try:
                segment_dict = segment.to_response_dict()
                segments_data.append(segment_dict)
                print(f'  ✅ Segmento {i+1} convertido: {segment_dict.get("_id", "sin_id")}')
            except Exception as e:
                print(f'  ❌ Error al convertir segmento {i+1}: {str(e)}')
        
        response = {
            'success': True,
            'message': 'Segmentos obtenidos exitosamente',
            'data': {
                'segments': segments_data,
                'count': len(segments_data),
                'project_id': project_id
            }
        }
        
        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)
        
    except Exception as error:
        print('💥 Error al obtener segmentos del proyecto:', str(error))
        import traceback
        print('📋 Stack trace:', traceback.format_exc())
        return jsonify({
            'success': False,
            'message': 'Error al obtener segmentos del proyecto'
        }), 500

def create_segment():
    """Crear un nuevo segmento"""
    try:
        data = request.get_json()
        start_time = data.get('startTime')
        end_time = data.get('endTime')
        project_id = data.get('projectid')
        prosody = data.get('prosody')
        prosody2 = data.get('prosody2')
        description = data.get('description')
        descriptions_prosody = data.get('Descriptions_prosody', [])
        
        print('📹 Creando nuevo segmento')
        print('📋 Datos recibidos:', {
            'start_time': start_time,
            'end_time': end_time,
            'project_id': project_id,
            'prosody': prosody,
            'prosody2': prosody2,
            'description': description,
            'descriptions_prosody': descriptions_prosody
        })

        # Validar datos requeridos
        if start_time is None or end_time is None or not project_id:
            return jsonify({
                'success': False,
                'message': 'startTime, endTime y projectid son requeridos'
            }), 400

        # Validar que los tiempos sean válidos
        if start_time < 0 or end_time < 0:
            return jsonify({
                'success': False,
                'message': 'Los tiempos deben ser mayores o iguales a 0'
            }), 400

        if start_time >= end_time:
            return jsonify({
                'success': False,
                'message': 'El tiempo de inicio debe ser menor al tiempo de fin'
            }), 400

        # Obtener base de datos
        db = get_db()

        # Verificar que el proyecto existe
        project = Project.find_by_id(db, project_id)
        if not project:
            print(f'❌ Proyecto no encontrado: {project_id}')
            return jsonify({
                'success': False,
                'message': 'Proyecto no encontrado'
            }), 404

        # Crear nuevo segmento
        segment = Segment(
            start_time=start_time,
            end_time=end_time,
            project_id=project_id,
            prosody=prosody,
            prosody2=prosody2,
            description=description,
            descriptions_prosody=descriptions_prosody
        )
        
        print('💾 Guardando segmento en la base de datos...')
        segment.save(db)
        print('✅ Segmento guardado exitosamente:', {
            '_id': str(segment._id),
            'start_time': segment.start_time,
            'end_time': segment.end_time,
            'project_id': segment.project_id
        })

        # Respuesta exitosa
        response = {
            'success': True,
            'message': 'Segmento creado exitosamente',
            'data': {
                'segment': segment.to_response_dict()
            }
        }

        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response), 201

    except Exception as error:
        print('💥 Error al crear segmento:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al crear segmento'
        }), 500

def update_segment(segment_id):
    """Actualizar un segmento"""
    try:
        data = request.get_json()
        start_time = data.get('startTime')
        end_time = data.get('endTime')
        prosody = data.get('prosody')
        prosody2 = data.get('prosody2')
        description = data.get('description')
        descriptions_prosody = data.get('Descriptions_prosody')
        
        print(f'📹 Actualizando segmento con ID: {segment_id}')
        print('📋 Datos recibidos:', {
            'start_time': start_time,
            'end_time': end_time,
            'prosody': prosody,
            'prosody2': prosody2,
            'description': description,
            'descriptions_prosody': descriptions_prosody
        })

        # Obtener base de datos
        db = get_db()

        # Buscar segmento
        segment = Segment.find_by_id(db, segment_id)
        
        if not segment:
            print(f'❌ Segmento no encontrado: {segment_id}')
            return jsonify({
                'success': False,
                'message': 'Segmento no encontrado'
            }), 404

        # Actualizar campos si se proporcionan
        if start_time is not None:
            if start_time < 0:
                return jsonify({
                    'success': False,
                    'message': 'El tiempo de inicio debe ser mayor o igual a 0'
                }), 400
            segment.start_time = start_time

        if end_time is not None:
            if end_time < 0:
                return jsonify({
                    'success': False,
                    'message': 'El tiempo de fin debe ser mayor o igual a 0'
                }), 400
            segment.end_time = end_time

        # Validar que los tiempos sean válidos juntos
        if start_time is not None or end_time is not None:
            if segment.start_time >= segment.end_time:
                return jsonify({
                    'success': False,
                    'message': 'El tiempo de inicio debe ser menor al tiempo de fin'
                }), 400

        if prosody is not None:
            segment.prosody = prosody
        if prosody2 is not None:
            segment.prosody2 = prosody2
        if description is not None:
            segment.description = description
        if descriptions_prosody is not None:
            segment.descriptions_prosody = descriptions_prosody
        
        print('💾 Guardando cambios en la base de datos...')
        segment.save(db)
        print('✅ Segmento actualizado exitosamente:', {
            '_id': str(segment._id),
            'start_time': segment.start_time,
            'end_time': segment.end_time
        })

        # Respuesta exitosa
        response = {
            'success': True,
            'message': 'Segmento actualizado exitosamente',
            'data': {
                'segment': segment.to_response_dict()
            }
        }

        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)

    except Exception as error:
        print('💥 Error al actualizar segmento:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al actualizar segmento'
        }), 500

def delete_segment(segment_id):
    """Eliminar un segmento"""
    try:
        print(f'📹 Eliminando segmento con ID: {segment_id}')
        
        # Obtener base de datos
        db = get_db()

        # Buscar segmento
        segment = Segment.find_by_id(db, segment_id)
        
        if not segment:
            print(f'❌ Segmento no encontrado: {segment_id}')
            return jsonify({
                'success': False,
                'message': 'Segmento no encontrado'
            }), 404

        # Eliminar segmento
        print('🗑️ Eliminando segmento de la base de datos...')
        segment.delete(db)
        print('✅ Segmento eliminado exitosamente:', segment_id)

        # Respuesta exitosa
        response = {
            'success': True,
            'message': 'Segmento eliminado exitosamente',
            'data': {
                'segment_id': segment_id
            }
        }

        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)

    except Exception as error:
        print('💥 Error al eliminar segmento:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al eliminar segmento'
        }), 500

def increment_views(segment_id):
    """Incrementar contador de vistas de un segmento"""
    try:
        print(f'📹 Incrementando vistas del segmento: {segment_id}')
        
        # Obtener base de datos
        db = get_db()

        # Buscar segmento
        segment = Segment.find_by_id(db, segment_id)
        
        if not segment:
            print(f'❌ Segmento no encontrado: {segment_id}')
            return jsonify({
                'success': False,
                'message': 'Segmento no encontrado'
            }), 404

        # Incrementar vistas
        segment.increment_views(db)
        print(f'✅ Vistas incrementadas: {segment.views}')

        # Respuesta exitosa
        response = {
            'success': True,
            'message': 'Vistas incrementadas exitosamente',
            'data': {
                'segment_id': segment_id,
                'views': segment.views
            }
        }

        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)

    except Exception as error:
        print('💥 Error al incrementar vistas:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al incrementar vistas'
        }), 500

def increment_likes(segment_id):
    """Incrementar contador de likes de un segmento"""
    try:
        print(f'📹 Incrementando likes del segmento: {segment_id}')
        
        # Obtener base de datos
        db = get_db()

        # Buscar segmento
        segment = Segment.find_by_id(db, segment_id)
        
        if not segment:
            print(f'❌ Segmento no encontrado: {segment_id}')
            return jsonify({
                'success': False,
                'message': 'Segmento no encontrado'
            }), 404

        # Incrementar likes
        segment.increment_likes(db)
        print(f'✅ Likes incrementados: {segment.likes}')

        # Respuesta exitosa
        response = {
            'success': True,
            'message': 'Likes incrementados exitosamente',
            'data': {
                'segment_id': segment_id,
                'likes': segment.likes
            }
        }

        print('📤 Enviando respuesta exitosa:', response)
        return jsonify(response)

    except Exception as error:
        print('💥 Error al incrementar likes:', str(error))
        return jsonify({
            'success': False,
            'message': 'Error al incrementar likes'
        }), 500 

File: controllers/segmentController.js
const Segment = require('../models/Segment');

/**
 * Obtener todos los segmentos con paginación y filtros
 * GET /api/segments
 */
const getAllSegments = async (req, res) => {
  try {
    const {
      page = 1,
      limit = 10,
      search,
      tags,
      sort = 'createdAt',
      order = 'desc',
      userId
    } = req.query;

    // Construir filtros
    const filters = { isPublic: true };
    
    if (search) {
      filters.$text = { $search: search };
    }
    
    if (tags) {
      const tagArray = tags.split(',').map(tag => tag.trim());
      filters.tags = { $in: tagArray };
    }
    
    if (userId) {
      filters.createdBy = userId;
    }

    // Construir opciones de consulta
    const options = {
      page: parseInt(page),
      limit: parseInt(limit),
      sort: { [sort]: order === 'desc' ? -1 : 1 },
      populate: {
        path: 'createdBy',
        select: 'name email'
      }
    };

    // Realizar consulta con paginación
    const segments = await Segment.paginate(filters, options);

    res.json({
      success: true,
      message: 'Segmentos obtenidos exitosamente',
      data: {
        segments: segments.docs,
        pagination: {
          page: segments.page,
          limit: segments.limit,
          totalDocs: segments.totalDocs,
          totalPages: segments.totalPages,
          hasNextPage: segments.hasNextPage,
          hasPrevPage: segments.hasPrevPage
        }
      }
    });

  } catch (error) {
    console.error('Error al obtener segmentos:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener segmentos',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Obtener segmento por ID
 * GET /api/segments/:id
 */
const getSegmentById = async (req, res) => {
  try {
    const segment = await Segment.findById(req.params.id)
      .populate('createdBy', 'name email');

    if (!segment) {
      return res.status(404).json({
        success: false,
        message: 'Segmento no encontrado'
      });
    }

    // Verificar si el usuario puede ver el segmento
    if (!segment.isPublic && (!req.user || segment.createdBy._id.toString() !== req.user._id.toString())) {
      return res.status(403).json({
        success: false,
        message: 'Acceso denegado'
      });
    }

    // Incrementar vistas si el usuario no es el creador
    if (req.user && segment.createdBy._id.toString() !== req.user._id.toString()) {
      await segment.incrementViews();
    }

    res.json({
      success: true,
      message: 'Segmento obtenido exitosamente',
      data: {
        segment
      }
    });

  } catch (error) {
    console.error('Error al obtener segmento:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener segmento',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Crear nuevo segmento
 * POST /api/segments
 */
const createSegment = async (req, res) => {
  try {
    const segmentData = {
      ...req.body,
      createdBy: req.user._id
    };

    const segment = new Segment(segmentData);
    await segment.save();

    // Poblar información del creador
    await segment.populate('createdBy', 'name email');

    res.status(201).json({
      success: true,
      message: 'Segmento creado exitosamente',
      data: {
        segment
      }
    });

  } catch (error) {
    console.error('Error al crear segmento:', error);
    res.status(500).json({
      success: false,
      message: 'Error al crear segmento',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Actualizar segmento
 * PUT /api/segments/:id
 */
const updateSegment = async (req, res) => {
  try {
    const segment = await Segment.findById(req.params.id);

    if (!segment) {
      return res.status(404).json({
        success: false,
        message: 'Segmento no encontrado'
      });
    }

    // Verificar permisos
    if (segment.createdBy.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
      return res.status(403).json({
        success: false,
        message: 'No tienes permisos para actualizar este segmento'
      });
    }

    // Actualizar segmento
    const updatedSegment = await Segment.findByIdAndUpdate(
      req.params.id,
      req.body,
      { new: true, runValidators: true }
    ).populate('createdBy', 'name email');

    res.json({
      success: true,
      message: 'Segmento actualizado exitosamente',
      data: {
        segment: updatedSegment
      }
    });

  } catch (error) {
    console.error('Error al actualizar segmento:', error);
    res.status(500).json({
      success: false,
      message: 'Error al actualizar segmento',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Eliminar segmento
 * DELETE /api/segments/:id
 */
const deleteSegment = async (req, res) => {
  try {
    const segment = await Segment.findById(req.params.id);

    if (!segment) {
      return res.status(404).json({
        success: false,
        message: 'Segmento no encontrado'
      });
    }

    // Verificar permisos
    if (segment.createdBy.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
      return res.status(403).json({
        success: false,
        message: 'No tienes permisos para eliminar este segmento'
      });
    }

    await Segment.findByIdAndDelete(req.params.id);

    res.json({
      success: true,
      message: 'Segmento eliminado exitosamente'
    });

  } catch (error) {
    console.error('Error al eliminar segmento:', error);
    res.status(500).json({
      success: false,
      message: 'Error al eliminar segmento',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Obtener segmentos del usuario actual
 * GET /api/segments/my-segments
 */
const getMySegments = async (req, res) => {
  try {
    const {
      page = 1,
      limit = 10,
      sort = 'createdAt',
      order = 'desc'
    } = req.query;

    const options = {
      page: parseInt(page),
      limit: parseInt(limit),
      sort: { [sort]: order === 'desc' ? -1 : 1 },
      populate: {
        path: 'createdBy',
        select: 'name email'
      }
    };

    const segments = await Segment.paginate(
      { createdBy: req.user._id },
      options
    );

    res.json({
      success: true,
      message: 'Mis segmentos obtenidos exitosamente',
      data: {
        segments: segments.docs,
        pagination: {
          page: segments.page,
          limit: segments.limit,
          totalDocs: segments.totalDocs,
          totalPages: segments.totalPages,
          hasNextPage: segments.hasNextPage,
          hasPrevPage: segments.hasPrevPage
        }
      }
    });

  } catch (error) {
    console.error('Error al obtener mis segmentos:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener mis segmentos',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Dar like a un segmento
 * POST /api/segments/:id/like
 */
const likeSegment = async (req, res) => {
  try {
    const segment = await Segment.findById(req.params.id);

    if (!segment) {
      return res.status(404).json({
        success: false,
        message: 'Segmento no encontrado'
      });
    }

    await segment.incrementLikes();

    res.json({
      success: true,
      message: 'Like agregado exitosamente',
      data: {
        likes: segment.likes
      }
    });

  } catch (error) {
    console.error('Error al dar like:', error);
    res.status(500).json({
      success: false,
      message: 'Error al dar like',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Quitar like de un segmento
 * POST /api/segments/:id/unlike
 */
const unlikeSegment = async (req, res) => {
  try {
    const segment = await Segment.findById(req.params.id);

    if (!segment) {
      return res.status(404).json({
        success: false,
        message: 'Segmento no encontrado'
      });
    }

    await segment.decrementLikes();

    res.json({
      success: true,
      message: 'Like removido exitosamente',
      data: {
        likes: segment.likes
      }
    });

  } catch (error) {
    console.error('Error al quitar like:', error);
    res.status(500).json({
      success: false,
      message: 'Error al quitar like',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Buscar segmentos por texto
 * GET /api/segments/search
 */
const searchSegments = async (req, res) => {
  try {
    const { q, page = 1, limit = 10 } = req.query;

    if (!q) {
      return res.status(400).json({
        success: false,
        message: 'Término de búsqueda requerido'
      });
    }

    const options = {
      page: parseInt(page),
      limit: parseInt(limit),
      sort: { score: { $meta: 'textScore' } },
      populate: {
        path: 'createdBy',
        select: 'name email'
      }
    };

    const segments = await Segment.paginate(
      { 
        $text: { $search: q },
        isPublic: true 
      },
      options
    );

    res.json({
      success: true,
      message: 'Búsqueda completada exitosamente',
      data: {
        segments: segments.docs,
        pagination: {
          page: segments.page,
          limit: segments.limit,
          totalDocs: segments.totalDocs,
          totalPages: segments.totalPages,
          hasNextPage: segments.hasNextPage,
          hasPrevPage: segments.hasPrevPage
        }
      }
    });

  } catch (error) {
    console.error('Error en búsqueda:', error);
    res.status(500).json({
      success: false,
      message: 'Error en búsqueda',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Obtener segmentos populares
 * GET /api/segments/popular
 */
const getPopularSegments = async (req, res) => {
  try {
    const { limit = 10 } = req.query;

    const segments = await Segment.find({ isPublic: true })
      .sort({ views: -1, likes: -1 })
      .limit(parseInt(limit))
      .populate('createdBy', 'name email');

    res.json({
      success: true,
      message: 'Segmentos populares obtenidos exitosamente',
      data: {
        segments
      }
    });

  } catch (error) {
    console.error('Error al obtener segmentos populares:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener segmentos populares',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

/**
 * Obtener segmentos recientes
 * GET /api/segments/recent
 */
const getRecentSegments = async (req, res) => {
  try {
    const { limit = 10 } = req.query;

    const segments = await Segment.find({ isPublic: true })
      .sort({ createdAt: -1 })
      .limit(parseInt(limit))
      .populate('createdBy', 'name email');

    res.json({
      success: true,
      message: 'Segmentos recientes obtenidos exitosamente',
      data: {
        segments
      }
    });

  } catch (error) {
    console.error('Error al obtener segmentos recientes:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener segmentos recientes',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
};

module.exports = {
  getAllSegments,
  getSegmentById,
  createSegment,
  updateSegment,
  deleteSegment,
  getMySegments,
  likeSegment,
  unlikeSegment,
  searchSegments,
  getPopularSegments,
  getRecentSegments
}; 